plugins/flatpak/gs-flatpak.c

```diff
@@ -1889,14 +1892,17 @@ gs_flatpak_refine_app_state_unlocked (GsFlatpak *self,
 		return FALSE;
 
 	/* find the app using the origin and the ID */
-	refs = flatpak_installation_list_installed_refs (self->installation,
-							 cancellable, error);
-	if (refs == NULL) {
+	if (self->installed_refs == NULL) {
+		self->installed_refs = flatpak_installation_list_installed_refs (self->installation,
```

Comment:
I think arch-wise/upstream-wise, this re-cache-ing bit go in :
1) gs_plugin_setup (i.e. initializing)
2) whenever the cache is invalidated (`gs_plugin_flatpak_changed_cb`)

It feels a bit weird to me to query the installed_refs while refining an app, so we can additionally improve on that too.
+1 on the approach overall though

Reply:
Thought about that but in the end I tried to keep the current approach of retrying to get the installed refs on refine as it may fail for some reason.